src/torchcodec/_core/Stream.h

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <torch/types.h>
+#include <optional>
+#include <string>
+
+namespace facebook::torchcodec {
+
+enum ColorConversionLibrary {
+  // TODO: Add an AUTO option later.
+  // Use the libavfilter library for color conversion.
+  FILTERGRAPH,
+  // Use the libswscale library for color conversion.
+  SWSCALE
+};
+
+struct VideoStreamOptions {
+  VideoStreamOptions() {}
+
+  // Number of threads we pass to FFMPEG for decoding.
+  // 0 means FFMPEG will choose the number of threads automatically to fully
+  // utilize all cores. If not set, it will be the default FFMPEG behavior for
+  // the given codec.
+  std::optional<int> ffmpegThreadCount;
+  // Currently the dimension order can be either NHWC or NCHW.
+  // H=height, W=width, C=channel.
+  std::string dimensionOrder = "NCHW";
+  // The output height and width of the frame. If not specified, the output
+  // is the same as the original video.
+  std::optional<int> width;
+  std::optional<int> height;
+  std::optional<ColorConversionLibrary> colorConversionLibrary;
+  // By default we use CPU for decoding for both C++ and python users.
+  torch::Device device = torch::kCPU;
+};
+
+struct AudioStreamOptions {
+  AudioStreamOptions() {}
+
+  std::optional<int> sampleRate;
+};
+
```

Comment:
Thanks for the PR @dvrogozh . Was it necessary to move the StreamMetadata and ContainerMetadata classes into this new header?
If yes, then I'd suggest moving them into their own `metadata.h` header
If not, then it might be best to leave them in the SingleStreamDecoder files

In both cases, I'd also suggest to rename Stream.h into just StreamOptions.h since this file should only contain option-related structs.


Otherwise, this PR LGTM

Reply:
In the last update I've pushed:
* `Stream.h` renamed to `StreamOptions.h`
* Created `Medata.h` to hold `StreamMetadata` and `ContainerMetadata`

@NicolasHug , @scotts : The reason why I moved metadata to header was that `StreamMetadata` is used by `FrameBatchOutput` which I also moved:

https://github.com/pytorch/torchcodec/blob/8b19f455046837f572490fe403918b57f7c1989a/src/torchcodec/_core/SingleStreamDecoder.h#L181-L184

That being said, note that I moved `ContainerMetadata` only for consistency - it's not used by any of the structures we really need to be in headers. Let me know if you want to avoid moving `ContainerMetadata`. Also, if we don't want to separately expose `FrameBatchOutput` we can avoid moving metadata to separate header entirely.

Reply:
@dvrogozh, looks good, thanks for cleaning this up!